### PR TITLE
Migrate from Zalando to Spring Framework 6 problem support implementation #19991[skip-ci]

### LINKED ISSUE #19991: Migrate from Zalando to Spring Framework 6 problem support implementation
##### **Overview of the issue**

Migrate from Zalando ([problem](https://github.com/zalando/problem) and [problem-spring-web](https://github.com/zalando/problem-spring-web)) to https://github.com/spring-projects/spring-framework/issues/27052 coming with spring-framework 6.0.

According to https://github.com/spring-projects/spring-boot/issues/32634, you might need to use a SNAPSHOT version (or 3.0.0-RC1 if it's released). 

##### **Motivation for or Use Case**

Zalando's implementation [doesn't support Spring Boot 3](https://github.com/zalando/problem-spring-web/issues/768). 

##### **Reproduce the error**

Create a new application using the `ngx-default` example's `.yo-rc.json`, then run `mvn test`.

```json
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "authenticationType": "jwt",
    "baseName": "jhipsterSampleApplication",
    "blueprints": [],
    "buildTool": "maven",
    "cacheProvider": "ehcache",
    "clientFramework": "angular",
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1596513172471,
    "cypressCoverage": true,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "devServerPort": 4200,
    "dtoSuffix": "",
    "embeddableLaunchScript": false,
    "enableGradleEnterprise": false,
    "enableHibernateCache": true,
    "enableTranslation": true,
    "entities": [],
    "entitySuffix": "Entity",
    "jhiPrefix": "myPrefix",
    "jhipsterVersion": "7.9.3",
    "jwtSecretKey": "ZjY4MTM4YjI5YzMwZjhjYjI2OTNkNTRjMWQ5Y2Q0Y2YwOWNmZTE2NzRmYzU3NTMwM2NjOTE3MTllOTM3MWRkMzcyYTljMjVmNmQ0Y2MxOTUzODc0MDhhMTlkMDIxMzI2YzQzZDM2ZDE3MmQ3NjVkODk3OTVmYzljYTQyZDNmMTQ=",
    "languages": [
      "en",
      "al",
      "ar-ly",
      "hy",
      "by",
      "bn",
      "bg",
      "ca",
      "zh-cn",
      "zh-tw",
      "cs",
      "da",
      "nl",
      "et",
      "fa",
      "fr",
      "gl",
      "de",
      "el",
      "hi",
      "hr",
      "hu",
      "id",
      "it",
      "ja",
      "ko",
      "mr",
      "my",
      "pl",
      "pt-br",
      "pt-pt",
      "ro",
      "ru",
      "si",
      "sk",
      "sr",
      "es",
      "sv",
      "tr",
      "ta",
      "te",
      "th",
      "ua",
      "uz-Cyrl-uz",
      "uz-Latn-uz",
      "vi"
    ],
    "messageBroker": false,
    "nativeLanguage": "en",
    "otherModules": [],
    "packageFolder": "tech/jhipster/sample",
    "packageName": "tech.jhipster.sample",
    "pages": [],
    "prodDatabaseType": "postgresql",
    "reactive": false,
    "searchEngine": false,
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipCheckLengthOfIdentifier": false,
    "skipClient": false,
    "skipFakeData": false,
    "skipUserManagement": false,
    "testFrameworks": ["gatling", "cypress"],
    "websocket": false,
    "withAdminUi": true
  }
}
```

##### **Related issues**

https://github.com/jhipster/generator-jhipster/issues/19782

### PR DESCRIPTION
**Pull request by dwarakaprasad**
fixes #19991
---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [X] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [X] Documentation is added/updated where necessary
- [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by dwarakaprasad** [comment]
@mraible this migration is complete except the [exception cause part](https://github.com/dwarakaprasad/generator-jhipster/blob/439fb7a2220c02fc5b535122d65fb3037c59cb77/generators/server/templates/src/main/java/package/web/rest/errors/ExceptionTranslator.java.ejs#L315), out of the box Spring's Problem implementation doesn't seem to support this (at least I couldn't find any documentation) like zalando. I can try something similar to zalando by [extending](https://github.com/spring-projects/spring-framework/issues/27052#issuecomment-1079248940) the base ProblemDetail class but have to look at how Spring handles the extended class during json conversion (with nesting)... 
What are your thoughts?

**Comment by mraible** [comment]
You might try creating an extended problem detail class and adding `cause` to it, like [this](https://github.com/rstoyanchev/sandbox-rfc7807/blob/main/src/main/java/org/springframework/sandbox/rfc7807/ExtendedProblemDetail.java) one does for `host`.

**Comment by dwarakaprasad** [comment]
> You might try creating an extended problem detail class and adding `cause` to it, like [this](https://github.com/rstoyanchev/sandbox-rfc7807/blob/main/src/main/java/org/springframework/sandbox/rfc7807/ExtendedProblemDetail.java) one does for `host`

Yeah sure. thanks

File: generators/server/templates/src/main/java/package/config/WebConfigurer.java.ejs [hunk 1]
```diff
@@ -215,13 +215,12 @@ public class WebConfigurer implements <% if (!reactive) { %>ServletContextInitia
     }
   <%_ } _%>
 
-  /*
     @Bean
     @Order(-2) // The handler must have precedence over WebFluxResponseStatusExceptionHandler and Spring Boot's ErrorWebExceptionHandler
```
Comment (DanielFran) [inline generators/server/templates/src/main/java/package/config/WebConfigurer.java.ejs:219]: Is it still needed?

Comment (dwarakaprasad) [reply generators/server/templates/src/main/java/package/config/WebConfigurer.java.ejs:219]: Yes, this is required as sometimes exceptions happen prior to the controller selection that doesn't end up being caught by the default or extended handlers. This PR introduces [ReactiveWebExceptionHandler](https://github.com/dwarakaprasad/jhipster-bom/blob/fix_19991_spring_boot3_problem/jhipster-framework/src/main/java/tech/jhipster/web/rest/errors/ReactiveWebExceptionHandler.java) that replaces zalando's ProblemExceptionHandler.

**Comment by mraible** [comment]
The CI failures don't seem to be related to this PR. I'm not sure what's causing them.

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project jhipster-sample-application: Compilation failure
Error:  /home/runner/work/generator-jhipster/generator-jhipster/app/src/main/java/tech/jhipster/sample/config/SecurityConfiguration.java:[49,17] method ignoringRequestMatchers in class org.springframework.security.config.annotation.web.configurers.CsrfConfigurer<H> cannot be applied to given types;
Error:    required: org.springframework.security.web.util.matcher.RequestMatcher[]
Error:    found:    java.lang.String
Error:    reason: varargs mismatch; java.lang.String cannot be converted to org.springframework.security.web.util.matcher.RequestMatcher
Error:  -> [Help 1]
```

**Comment by dwarakaprasad** [comment]
> The CI failures don't seem to be related to this PR. I'm not sure what's causing them.
> 
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project jhipster-sample-application: Compilation failure
> Error:  /home/runner/work/generator-jhipster/generator-jhipster/app/src/main/java/tech/jhipster/sample/config/SecurityConfiguration.java:[49,17] method ignoringRequestMatchers in class org.springframework.security.config.annotation.web.configurers.CsrfConfigurer<H> cannot be applied to given types;
> Error:    required: org.springframework.security.web.util.matcher.RequestMatcher[]
> Error:    found:    java.lang.String
> Error:    reason: varargs mismatch; java.lang.String cannot be converted to org.springframework.security.web.util.matcher.RequestMatcher
> Error:  -> [Help 1]
> ```

I have fixed it along with 3 other test failures. Will be pushing the changes in a bit

File: .github/actions/setup/action.yml [hunk 2]
```diff
@@ -63,11 +63,11 @@ inputs:
   jhipster-bom-repository:
     description: 'JHipster BOM repository'
     required: false
-    default: https://github.com/jhipster/jhipster-bom.git
+    default: https://github.com/dwarakaprasad/jhipster-bom
```
Comment (mshima) [inline .github/actions/setup/action.yml:66]:
```suggestion
    default: https://github.com/jhipster/jhipster-bom.git
```

File: .github/actions/setup/action.yml [hunk 3]
```diff
@@ -63,11 +63,11 @@ inputs:
...
   jhipster-bom-branch:
     description: 'JHipster BOM branch'
     required: false
-    default: spring-boot-3.0-m4
+    default: fix_19991_spring_boot3_problem
```
Comment (mshima) [inline .github/actions/setup/action.yml:70]:
```suggestion
    default: spring-boot-3.0-m4
```

**Review by mshima — APPROVED** [review]
Great work.

**Comment by dwarakaprasad** [comment]
$500 bug bounty claim [https://opencollective.com/generator-jhipster/expenses/160399](https://opencollective.com/generator-jhipster/expenses/160399)

**Pull request merged**